06_HelloGraphicsQueue/main.cpp

```diff
@@ -2,80 +2,436 @@
+		// We will cover allocators and suballocation in a later example.
+		nbl::video::IDeviceMemoryAllocator::SMemoryOffset outputBufferAllocation = {};
+		smart_refctd_ptr<IGPUBuffer> outputImageBuffer = nullptr;
 		{
-			// Remember to call the base class initialization!
-			if (!device_base_t::onAppInitialized(std::move(system)))
-				return false;
-			if (!asset_base_t::onAppInitialized(std::move(system)))
-				return false;
+			IGPUBuffer::SCreationParams gpuBufCreationParams;
+			gpuBufCreationParams.size = smallImgByteSize;
+			// VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT
+			gpuBufCreationParams.usage = IGPUBuffer::E_USAGE_FLAGS::EUF_TRANSFER_DST_BIT;
+			outputImageBuffer = m_device->createBuffer(std::move(gpuBufCreationParams));
+			if (!outputImageBuffer)
+				return logFail("Failed to create a GPU Buffer of size %d!\n", gpuBufCreationParams.size);
+
+			// Naming objects is cool because not only errors (such as Vulkan Validation Layers) will show their names, but RenderDoc captures too.
+			outputImageBuffer->setObjectDebugName("Output Image Buffer");
+
+			nbl::video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = outputImageBuffer->getMemoryReqs();
+			// you can simply constrain the memory requirements by AND-ing the type bits of the host visible memory types
+			reqs.memoryTypeBits &= m_physicalDevice->getHostVisibleMemoryTypeBits();
+
+			// There are actually two `allocate` overloads, one which allocates memory if you already know the type you want.
+			// And this one which is a utility which tries to allocate from every type that matches your requirements in some order of preference.
+			// The other of preference (iteration over compatible types) can be controlled by the method's template parameter,
+			// the default is from lowest index to highest, but skipping over incompatible types.
+			outputBufferAllocation = m_device->allocate(reqs, outputImageBuffer.get(), nbl::video::IDeviceMemoryAllocation::E_MEMORY_ALLOCATE_FLAGS::EMAF_NONE);
+			if (!outputBufferAllocation.isValid())
+				return logFail("Failed to allocate Device Memory compatible with our GPU Buffer!\n");
+
+			// Note that we performed a Dedicated Allocation above, so there's no need to bind the memory anymore (since the allocator knows the dedication, it can already bind).
+			// This is a carryover from having an OpenGL backend, where you couldn't have a memory allocation separate from the resource, so all allocations had to be "dedicated".
+			// In Vulkan dedicated allocations are the most performant and still make sense as long as you won't blow the 4096 allocation limit on windows.
+			// You should always use dedicated allocations for images used for swapchains, framebuffer attachments (esp transient), as well as objects used in CUDA/DirectX interop.
+			assert(outputImageBuffer->getBoundMemory() == outputBufferAllocation.memory.get());
+		}
```

Comment:
you don't need to copy comments from previous examples, only explain things that are new